File: tests/algorithms/x/Java/maths/quadratic_equations_complex_numbers.java
public class Main {
    static class Complex {
        double re;
        double im;
        Complex(double re, double im) {
            this.re = re;
            this.im = im;
        }
        Complex() {}
        @Override public String toString() {
            return String.format("{'re': %s, 'im': %s}", String.valueOf(re), String.valueOf(im));
        }
    }


    static Complex add(Complex a, Complex b) {
        return new Complex((double)(a.re) + (double)(b.re), (double)(a.im) + (double)(b.im));
    }

    static Complex sub(Complex a, Complex b) {
        return new Complex((double)(a.re) - (double)(b.re), (double)(a.im) - (double)(b.im));
    }

    static Complex div_real(Complex a, double r) {
        return new Complex((double)(a.re) / (double)(r), (double)(a.im) / (double)(r));
    }

    static double sqrt_newton(double x) {
        if ((double)(x) <= (double)(0.0)) {
            return 0.0;
        }
        double guess_1 = (double)((double)(x) / (double)(2.0));
        long i_1 = 0L;
        while ((long)(i_1) < 20L) {
            guess_1 = (double)((double)(((double)(guess_1) + (double)((double)(x) / (double)(guess_1)))) / (double)(2.0));
            i_1 = (long)((long)(i_1) + 1L);
        }
        return guess_1;
    }

    static Complex sqrt_to_complex(double d) {
        if ((double)(d) >= (double)(0.0)) {
            return new Complex(sqrt_newton((double)(d)), 0.0);
        }
        return new Complex(0.0, sqrt_newton((double)(-d)));
    }

    static Complex[] quadratic_roots(double a, double b, double c) {
        if ((double)(a) == (double)(0.0)) {
            System.out.println("ValueError: coefficient 'a' must not be zero");
            return new Complex[]{};
        }
        double delta_1 = (double)((double)((double)(b) * (double)(b)) - (double)((double)((double)(4.0) * (double)(a)) * (double)(c)));
        Complex sqrt_d_1 = sqrt_to_complex((double)(delta_1));
        Complex minus_b_1 = new Complex(-b, 0.0);
        double two_a_1 = (double)((double)(2.0) * (double)(a));
        Complex root1_1 = div_real(add(minus_b_1, sqrt_d_1), (double)(two_a_1));
        Complex root2_1 = div_real(sub(minus_b_1, sqrt_d_1), (double)(two_a_1));
        return new Complex[]{root1_1, root2_1};
    }

    static String root_str(Complex r) {
        if ((double)(r.im) == (double)(0.0)) {
            return _p(r.re);
        }
        String s_1 = _p(r.re);
        if ((double)(r.im) >= (double)(0.0)) {
            s_1 = s_1 + "+" + _p(r.im) + "i";
        } else {
            s_1 = s_1 + _p(r.im) + "i";
        }
        return s_1;
    }

    static void main() {
        Complex[] roots = ((Complex[])(quadratic_roots((double)(5.0), (double)(6.0), (double)(1.0))));
        if ((long)(roots.length) == 2L) {
            System.out.println("The solutions are: " + String.valueOf(root_str(roots[(int)(0L)])) + " and " + String.valueOf(root_str(roots[(int)(1L)])));
        }
    }
    public static void main(String[] args) {
        main();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/transpiler/FS/maths/quadratic_equations_complex_numbers.fs
// Generated 2025-08-17 12:28 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.10g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
type Complex = {
    mutable _re: float
    mutable _im: float
}
let rec add (a: Complex) (b: Complex) =
    let mutable __ret : Complex = Unchecked.defaultof<Complex>
    let mutable a = a
    let mutable b = b
    try
        __ret <- { _re = (a._re) + (b._re); _im = (a._im) + (b._im) }
        raise Return
        __ret
    with
        | Return -> __ret
and sub (a: Complex) (b: Complex) =
    let mutable __ret : Complex = Unchecked.defaultof<Complex>
    let mutable a = a
    let mutable b = b
    try
        __ret <- { _re = (a._re) - (b._re); _im = (a._im) - (b._im) }
        raise Return
        __ret
    with
        | Return -> __ret
and div_real (a: Complex) (r: float) =
    let mutable __ret : Complex = Unchecked.defaultof<Complex>
    let mutable a = a
    let mutable r = r
    try
        __ret <- { _re = (a._re) / r; _im = (a._im) / r }
        raise Return
        __ret
    with
        | Return -> __ret
and sqrt_newton (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        if x <= 0.0 then
            __ret <- 0.0
            raise Return
        let mutable guess: float = x / 2.0
        let mutable i: int = 0
        while i < 20 do
            guess <- (guess + (x / guess)) / 2.0
            i <- i + 1
        __ret <- guess
        raise Return
        __ret
    with
        | Return -> __ret
and sqrt_to_complex (d: float) =
    let mutable __ret : Complex = Unchecked.defaultof<Complex>
    let mutable d = d
    try
        __ret <- if d >= 0.0 then { _re = sqrt_newton (d); _im = 0.0 } else { _re = 0.0; _im = sqrt_newton (-d) }
        raise Return
        __ret
    with
        | Return -> __ret
and quadratic_roots (a: float) (b: float) (c: float) =
    let mutable __ret : Complex array = Unchecked.defaultof<Complex array>
    let mutable a = a
    let mutable b = b
    let mutable c = c
    try
        if a = 0.0 then
            ignore (printfn "%s" ("ValueError: coefficient 'a' must not be zero"))
            __ret <- Array.empty<Complex>
            raise Return
        let delta: float = (b * b) - ((4.0 * a) * c)
        let sqrt_d: Complex = sqrt_to_complex (delta)
        let minus_b: Complex = { _re = -b; _im = 0.0 }
        let two_a: float = 2.0 * a
        let root1: Complex = div_real (add (minus_b) (sqrt_d)) (two_a)
        let root2: Complex = div_real (sub (minus_b) (sqrt_d)) (two_a)
        __ret <- unbox<Complex array> [|root1; root2|]
        raise Return
        __ret
    with
        | Return -> __ret
and root_str (r: Complex) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable r = r
    try
        if (r._im) = 0.0 then
            __ret <- _str (r._re)
            raise Return
        let mutable s: string = _str (r._re)
        if (r._im) >= 0.0 then
            s <- ((s + "+") + (_str (r._im))) + "i"
        else
            s <- (s + (_str (r._im))) + "i"
        __ret <- s
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let roots: Complex array = quadratic_roots (5.0) (6.0) (1.0)
        if (Seq.length (roots)) = 2 then
            ignore (printfn "%s" ((("The solutions are: " + (root_str (_idx roots (int 0)))) + " and ") + (root_str (_idx roots (int 1)))))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
ignore (main())
